src/decode/delta.rs

```diff
@@ -0,0 +1,65 @@
+use crate::decode::lzbuffer;
+use crate::decode::lzbuffer::LzBuffer;
+use crate::error;
+use byteorder::ReadBytesExt;
+use std::io;
+
+#[derive(Debug)]
+/// Decoder for XZ delta-encoded blocks (filter 3).
+pub struct DeltaDecoder {
+    distance: usize,
+    pos: u8,
+    delta: [u8; 256],
+}
+
+impl DeltaDecoder {
+    /// Creates a new object ready for transforming data that it's given.
+    pub fn new(property_distance: u8) -> Self {
+        DeltaDecoder {
+            distance: property_distance as usize + 1,
+            pos: 0,
+            delta: [0u8; 256],
+        }
+    }
+
+    /// Performs the equivalent of replacing this decompression state with a
+    /// freshly allocated copy.
+    ///
+    /// This function may not allocate memory and will attempt to reuse any
+    /// previously allocated resources.
+    #[cfg(feature = "raw_decoder")]
+    pub fn reset(&mut self) {
+        self.pos = 0;
+        self.delta = [0u8; 256];
+    }
+
+    /// Decompresses the input data into the output, consuming only as much
+    /// input as needed and writing as much output as possible.
+    pub fn decompress<W: io::Write, R: io::BufRead>(
+        &mut self,
+        input: &mut R,
+        output: &mut W,
+    ) -> error::Result<()> {
+        let mut accum = lzbuffer::LzAccumBuffer::from_stream(output, usize::MAX);
+
+        // See xz-file-format.txt for the C pseudocode this is implementing.
+        loop {
+            let byte = if let Ok(byte) = input.read_u8() {
+                byte
+            } else {
+                lzma_info!("Delta end of input");
+                break;
+            };
+
+            let tmp = self.delta[(self.distance + self.pos as usize) as u8 as usize];
```

Comment:
The original distance was a `u8`, so yes wrapping arithmetic on `u8` should work (if `self.distance` is made a `u8`) :) Using `self.pos.wrapping_sub(self.distance)` would be more readable though (with `pos` growing rather than decreasing).

Reply:
Good observation! You are quite correct. I had assumed that because we're assigning deltas back into the buffer the iteration order would matter, but in fact we only need to make sure that the relative distance between pos and distance are correct and the consistent.